### ISSUE TITLE
pam_access: IPv6 in `/etc/security/access.conf` requires interface to be specified since 1.5.3

### ISSUE THREAD

**Issue by charlesmaher** [linux-pam/linux-pam#582]
This is a breaking change that I can't find mentioned anywhere

Example `access.conf`:
```r
+ : ALL : LOCAL
+ : ALL : 192.168.0.0/24

# Works in 1.5.2, not in 1.5.3
+ : ALL : fe80::/10
+ : ALL : fe80::a1:b2c3

# Works in 1.5.2 AND 1.5.3
+ : ALL : fe80::a1:b2c3%eth0
# but I can't figure out how to specify an address range with interface

- : ALL : ALL
```

Here's what my current setup on Arch Linux looks like.

<details>

`/etc/pam.d/system-remote-login`:

```r
#%PAM-1.0

# only remote users need TOTP
auth      [success=1 default=ignore]  pam_access.so  accessfile=/etc/security/access-local.conf debug
auth      required                    pam_oath.so    usersfile=/etc/users.oath window=30 digits=6

auth      include                     system-login
account   include                     system-login
password  include                     system-login
session   include                     system-login
```

`/etc/security/access-local.conf`:

```r
# only allow inside local network
+ : ALL : LOCAL fe80::/10 192.168.0.0/24

# must uncomment this to avoid entering TOTP with version 1.5.3
#+ : user0 : fe80::a1:b2c3%eth0

- : ALL : ALL
```

</details>

And here are the logs:

<details>
<summary>Version 1.5.2</summary>

### With `pam-1.5.2-2-x86_64.pkg.tar.zst`

```r
pam_access(sshd:auth): login_access: user=user0, from=fe80::a1:b2c3%eth0, file=/etc/security/access-local.conf
pam_access(sshd:auth): line 2: +  :  ALL  :  LOCAL fe80::/10 192.168.0.0/24
pam_access(sshd:auth): list_match: list= ALL , item=user0
pam_access(sshd:auth): user_match: tok=ALL, item=user0
pam_access(sshd:auth): string_match: tok=ALL, item=user0
pam_access(sshd:auth): user_match=2, "user0"
pam_access(sshd:auth): list_match: list= LOCAL fe80::/10 192.168.0.0/24, item=user0
pam_access(sshd:auth): from_match: tok=LOCAL, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): string_match: tok=LOCAL, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): from_match: tok=fe80::/10, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): string_match: tok=fe80::/10, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): network_netmask_match: tok=fe80::/10, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): from_match=1, "fe80::a1:b2c3%eth0"
Accepted keyboard-interactive/pam for user0 from fe80::a1:b2c3%eth0 port 23800 ssh2
pam_unix(sshd:session): session opened for user user0(uid=1000) by (uid=0)
```
</details>

<details>
<summary>Version 1.5.3</summary>

### With `pam-1.5.3-1-x86_64.pkg.tar.zst`

```r
pam_access(sshd:auth): login_access: user=user0, from=fe80::a1:b2c3%eth0, file=/etc/security/access-local.conf
pam_access(sshd:auth): line 2: +  :  ALL  :  LOCAL fe80::/10 192.168.0.0/24
pam_access(sshd:auth): list_match: list= ALL , item=user0
pam_access(sshd:auth): user_match: tok=ALL, item=user0
pam_access(sshd:auth): string_match: tok=ALL, item=user0
pam_access(sshd:auth): user_match=2, "user0"
pam_access(sshd:auth): list_match: list= LOCAL fe80::/10 192.168.0.0/24, item=user0
pam_access(sshd:auth): from_match: tok=LOCAL, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): string_match: tok=LOCAL, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): from_match: tok=fe80::/10, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): string_match: tok=fe80::/10, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): network_netmask_match: tok=fe80::/10, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): from_match: tok=192.168.0.0/24, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): string_match: tok=192.168.0.0/24, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): network_netmask_match: tok=192.168.0.0/24, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): from_match=0, "fe80::a1:b2c3%eth0"
pam_access(sshd:auth): line 7: -  :  ALL  :  ALL
pam_access(sshd:auth): list_match: list= ALL , item=user0
pam_access(sshd:auth): user_match: tok=ALL, item=user0
pam_access(sshd:auth): string_match: tok=ALL, item=user0
pam_access(sshd:auth): user_match=2, "user0"
pam_access(sshd:auth): list_match: list= ALL, item=user0
pam_access(sshd:auth): from_match: tok=ALL, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): string_match: tok=ALL, item=fe80::a1:b2c3%eth0
pam_access(sshd:auth): from_match=2, "fe80::a1:b2c3%eth0"
pam_access(sshd:auth): access denied for user `user0' from `fe80::a1:b2c3%eth0`

# After typing TOTP code
Accepted keyboard-interactive/pam for user0 from fe80::a1:b2c3%eth0 port 23804 ssh2
pam_unix(sshd:session): session opened for user user0(uid=1000) by (uid=0)
```
</details>

**Comment by ldv-alt** [comment]
Could it be an unexpected side effect of commit 23393bef92c1e768eda329813d7af55481c6ca9f?

**Comment by charlesmaher** [comment]
> Could it be an unexpected side effect of commit [23393be](https://github.com/linux-pam/linux-pam/commit/23393bef92c1e768eda329813d7af55481c6ca9f)?

Yep, that's it! I compiled and installed the commit prior to that one (9e800c2) in a VM and there was no issue, then I installed
23393be and the issue reappeared.

**Comment by ldv-alt** [comment]
@charlesmaher, thanks for confirming.
@thkukuk, could you have a look, please?

**Comment by thkukuk** [comment]
fe80:: is a link local IPv6 address, so a very special case. And if you read this the result is not unexpected: https://www.itdojo.com/ssh-using-link-local-ipv6-addresses/

"Normal" IPv6 addresses work fine.

I just don't understand yet why it worked before, and how to "fix" it correctly if it is a bug at all. Because fe80::a1:b2c3 must not be the same as fe80::a1:b2c3, as it is link-local and not machine local.

**Comment by charlesmaher** [comment]
Ah, you're right! The issue is that I was using `fe80::`. In that case I certainly don't mind specifying the interface, but this will make the [current documentation](https://github.com/linux-pam/linux-pam/blob/1fbf123d982b90d41463df7b6b59a4e544263358/modules/pam_access/access.conf.5.xml#L194) out of date since it will no longer work for "all" addresses.
Explaining the syntax for link-local addresses and ranges would be enough for me, though it is a shame that some existing configurations will be invalid.

**Comment by thkukuk** [comment]
@ldv-alt @t8m
IPv6 with link-local addresses is a mess, depending on the function and protocol version of that function it works or does not work. It looks like with 1.5.2 and older it works by accident (inet_ntop does not know about %<interface> and thus does not write it into the buffer), with 1.5.3 we run into the issue, that getnameinfo() function is used, which does understands the required "%<interface>" semantic in newer implementations, but our `isipaddr()` function does not understand it and thus `are_addresses_equal()` fails.

I see the following options:
1. ignore the interface like in the past and remove it at the beginning of the `network_netmask_match()` function
2. ignore the interface only in `isipaddr()`
3. document, that with link-local IPv6 addresses, only full IPv6 addresses with interface works
4. find an IPv6 expert, which could implement the comparison of two IPv6 addresses correct.

What do you think?

**Comment by thkukuk** [comment]
After discussing with some people, I would tend to option 3.

**Comment by charlesmaher** [comment]
Option 3 is good, but how should I match on `fe80::` addresses? Neither `fe80::/10%eth0` nor `fe80::%eth0/10` work on my machine.
At the moment my only options are to disable IPv6 or hard-code all the addresses I need in `/etc/security/access.conf`.

**Comment by thkukuk** [comment]
> Option 3 is good, but how should I match on `fe80::` addresses? Neither `fe80::/10%eth0` nor `fe80::%eth0/10` work on my machine. At the moment my only options are to disable IPv6 or hard-code all the addresses I need in `/etc/security/access.conf`.

Or you don't use link local addresses...

According to RFC 4007, the handling of link local addresses until Linux-PAM 1.5.2 is wrong and a security risk.
With 1.5.3, depending on the age of your getnameinfo() implementation, we face the same problems, with too old implementations, we have a security problem, with newer implementations it only works with addresses, not prefixes.
Until a IPv4/IPv6 experts steps up and re-implements the whole stuff in pam_access, we have to life either with an insecure implementation (which means, the advice can only be to never use link local addresses in access.conf) or with the current limitation for link local addresses.

**Comment by t8m** [comment]
I think documenting the current behavior is a sufficient fix.

**Comment by ldv-alt** [comment]
Addressed by #594
